packages/puppeteer-core/src/bidi/CdpSession.ts

```diff
@@ -0,0 +1,103 @@
+import type ProtocolMapping from 'devtools-protocol/types/protocol-mapping.js';
+
+import {CDPSession} from '../api/CDPSession.js';
+import type {Connection as CdpConnection} from '../cdp/Connection.js';
+import {TargetCloseError, UnsupportedOperation} from '../common/Errors.js';
+import {assert} from '../util/assert.js';
+import {Deferred} from '../util/Deferred.js';
+
+import type {BidiConnection} from './Connection.js';
+import type {BidiFrame} from './Frame.js';
+
+/**
+ * @internal
+ */
+export class BidiCdpSession extends CDPSession {
+  static sessions = new Map<string, BidiCdpSession>();
+
+  #frame: BidiFrame;
+  #sessionId = Deferred.create<string>();
+  #detached = false;
+
+  constructor(frame: BidiFrame, sessionId?: string) {
+    super();
+    this.#frame = frame;
+    if (sessionId !== undefined) {
+      this.#sessionId.resolve(sessionId);
+    }
+
+    void this.#initialize();
+  }
+
+  get #connection(): BidiConnection {
+    return this.#frame.browserContext.browser().session
+      .connection as BidiConnection;
+  }
+
+  get #supported() {
+    return this.#frame.browserContext.browser().cdpSupported;
+  }
+
+  async #initialize() {
+    if (!this.#supported) {
+      return;
+    }
+    if (!this.#sessionId.finished()) {
+      const {
+        result: {session},
+      } = await this.#connection.send('cdp.getSession', {
+        context: this.#frame._id,
+      });
+      assert(session);
+      this.#sessionId.resolve(session);
+    }
+
+    // SAFETY: We never throw #sessionId.
+    BidiCdpSession.sessions.set(this.#sessionId.value() as string, this);
+  }
+
+  override connection(): CdpConnection | undefined {
+    return undefined;
+  }
+
+  override async send<T extends keyof ProtocolMapping.Commands>(
+    method: T,
+    ...paramArgs: ProtocolMapping.Commands[T]['paramsType']
```

Comment:
Non blocker: Please add `TODO:` we should support the same behavior as CDP implementation to have timeout here.